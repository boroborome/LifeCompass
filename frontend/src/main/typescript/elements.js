// import only used components to reduce code size
export { default as Button } from "element-ui/lib/button";
export { default as ButtonGroup } from "element-ui/lib/button-group";
export { default as Col } from "element-ui/lib/col";
export { default as MessageBox } from "element-ui/lib/message-box";
export { default as Option } from "element-ui/lib/option";
export { default as Row } from "element-ui/lib/row";
export { default as Select } from "element-ui/lib/select";
export { default as TabPane } from "element-ui/lib/tab-pane";
export { default as Tabs } from "element-ui/lib/tabs";

import locale from "element-ui/lib/locale";
import de from "element-ui/lib/locale/lang/de";
import en from "element-ui/lib/locale/lang/en";

export function configureLocale(lcl) {
    locale.use(/^de($|-)/.test(lcl) ? de : en);
}
