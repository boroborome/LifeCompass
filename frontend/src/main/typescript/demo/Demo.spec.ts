import {TaskDto} from "@Generated/openapi";
import {config, mount} from "@vue/test-utils";
import {createRenderer} from "vue-server-renderer";
import {MessageBox} from "../elements";
import {reactive} from "../reactive";
import {Suggestions} from "../state/suggestions";
import {ConflictError, TaskList} from "../state/taskList";
import Demo from "./Demo.vue";

config.logModifiedComponents = false;

const renderer = createRenderer();

const mockConfirm = MessageBox.confirm = jest.fn();
const mockAdd = jest.fn();
const mockRemove = jest.fn();
const mockRefresh = jest.fn();
const mockReset = jest.fn();
const mockSave = jest.fn();

@reactive
class TaskListFake {
    constructor(public tasks: TaskDto[] = [], public changed = false) {
    }

    add(...args: any[]) {
        mockAdd(...args);
    }

    remove(...args: any[]) {
        mockRemove(...args);
    }

    async refresh() {
        await mockRefresh();
    }

    reset() {
        mockReset();
    }

    async save() {
        await mockSave();
    }
}

const taskListFake = new TaskListFake();

afterEach(() => {
    mockConfirm.mockReset();
    mockAdd.mockReset();
    mockRemove.mockReset();
    mockRefresh.mockReset();
    mockReset.mockReset();
    mockSave.mockReset();
    taskListFake.tasks = [];
    taskListFake.changed = false;
});

function demo() {
    return mount(Demo, {
        mocks: {
            $t: (key: string) => key,
            $tc: (key: string) => key,
        },
        provide: {
            [Suggestions.name]: {
                openSuggestions: [],
            },
            [TaskList.name]: taskListFake,
        },
    }).vm;
}

describe("tasklist app", () => {
    it("is initially invalid", () => {
        expect(demo().valid).toBe(false);
    });

    it("becomes valid for sufficient long task name", () => {
        const vm = demo();
        vm.newTask = "1234";
        expect(vm.valid).toBe(true);
    });

    it("adds valid new task", () => {
        const vm = demo();
        vm.newTask = "1234";
        vm.addNewTask();
        expect(mockAdd).toBeCalledWith("1234");
        expect(vm.newTask).toBe("");
    });

    it("ignores invalid new task", () => {
        const vm = demo();
        vm.newTask = "123";
        vm.addNewTask();
        expect(mockAdd).not.toBeCalled();
        expect(vm.newTask).toBe("123");
    });

    it("closes task", async () => {
        const vm = demo();
        taskListFake.tasks.push({title: "XXX", completed: false});
        vm.close(taskListFake.tasks[0]);
        expect(await renderer.renderToString(vm)).toMatchSnapshot();
    });

    it("opens task", async () => {
        const vm = demo();
        taskListFake.tasks.push({title: "XXX", completed: true});
        vm.open(taskListFake.tasks[0]);
        expect(await renderer.renderToString(vm)).toMatchSnapshot();
    });

    it("closes all tasks", async () => {
        const vm = demo();
        taskListFake.tasks.push({title: "XXX", completed: false});
        vm.closeAll();
        expect(await renderer.renderToString(vm)).toMatchSnapshot();
    });

    it("resets task list", () => {
        demo().reset();
        expect(mockReset).toBeCalled();
    });

    it("removes task", () => {
        const task = {title: "task", completed: true};
        demo().remove(task);
        expect(mockRemove).toBeCalledWith(task);
    });

    it("removes error class on valid input", async () => {
        const vm = demo();
        expect(await renderer.renderToString(vm)).toMatchSnapshot();
        vm.newTask = "1234";
        expect(await renderer.renderToString(vm)).toMatchSnapshot();
    });

    it("saves task list", async () => {
        await demo().save();
        expect(mockSave).toBeCalled();
    });

    it("handles unknow error on save", async () => {
        const error = {};
        mockSave.mockRejectedValue(error);
        await expect(demo().save()).rejects.toBe(error);
    });

    it("refreshes on conflict error", async () => {
        mockSave.mockRejectedValue(new ConflictError());
        await demo().save();
        expect(mockRefresh).toBeCalled();
    });

    it("does not refresh on conflict error", async () => {
        mockSave.mockRejectedValue(new ConflictError());
        mockConfirm.mockRejectedValue({});
        await demo().save();
        expect(mockRefresh).not.toBeCalled();
    });
});
