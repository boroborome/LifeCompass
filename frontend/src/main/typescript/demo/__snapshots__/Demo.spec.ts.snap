// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tasklist app closes all tasks 1`] = `
<div>
  <div class="el-row">
    <div class="el-col el-col-12">
      <div class="tasklist">
        <h1>tasks</h1>
        <div class="el-input input"><input type="text" placeholder="addTask" value="" class="el-input__inner error"></div>
        <div class="el-button-group"><button disabled="disabled" type="button" class="el-button el-button--primary is-disabled">
            <!---->
            <!----><span>markAllAsDone</span></button> <button disabled="disabled" type="button" class="el-button el-button--primary is-disabled">
            <!---->
            <!----><span>reset</span></button> <button disabled="disabled" type="button" class="el-button el-button--primary is-disabled">
            <!---->
            <!----><span>save</span></button></div>
        <hr>
        <ul class="items"></ul>
        <div class="footer">itemsLeft</div>
      </div>
    </div>
    <div class="el-col el-col-12">
      <div class="tasklist">
        <h1>alreadyDone</h1>
        <ul class="items">
          <li class="done">
            <div class="el-row is-justify-space-between el-row--flex">
              <div class="el-col el-col-20">XXX</div>
              <div class="el-col el-col-4">
                <div class="el-button-group"><button type="button" class="el-button el-button--default el-button--mini">
                    <!----><i class="el-icon-back"></i>
                    <!----></button> <button type="button" class="el-button el-button--default el-button--mini">
                    <!----><i class="el-icon-close"></i>
                    <!----></button></div>
              </div>
            </div>
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;

exports[`tasklist app closes task 1`] = `
<div>
  <div class="el-row">
    <div class="el-col el-col-12">
      <div class="tasklist">
        <h1>tasks</h1>
        <div class="el-input input"><input type="text" placeholder="addTask" value="" class="el-input__inner error"></div>
        <div class="el-button-group"><button disabled="disabled" type="button" class="el-button el-button--primary is-disabled">
            <!---->
            <!----><span>markAllAsDone</span></button> <button disabled="disabled" type="button" class="el-button el-button--primary is-disabled">
            <!---->
            <!----><span>reset</span></button> <button disabled="disabled" type="button" class="el-button el-button--primary is-disabled">
            <!---->
            <!----><span>save</span></button></div>
        <hr>
        <ul class="items"></ul>
        <div class="footer">itemsLeft</div>
      </div>
    </div>
    <div class="el-col el-col-12">
      <div class="tasklist">
        <h1>alreadyDone</h1>
        <ul class="items">
          <li class="done">
            <div class="el-row is-justify-space-between el-row--flex">
              <div class="el-col el-col-20">XXX</div>
              <div class="el-col el-col-4">
                <div class="el-button-group"><button type="button" class="el-button el-button--default el-button--mini">
                    <!----><i class="el-icon-back"></i>
                    <!----></button> <button type="button" class="el-button el-button--default el-button--mini">
                    <!----><i class="el-icon-close"></i>
                    <!----></button></div>
              </div>
            </div>
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;

exports[`tasklist app opens task 1`] = `
<div>
  <div class="el-row">
    <div class="el-col el-col-12">
      <div class="tasklist">
        <h1>tasks</h1>
        <div class="el-input input"><input type="text" placeholder="addTask" value="" class="el-input__inner error"></div>
        <div class="el-button-group"><button type="button" class="el-button el-button--primary">
            <!---->
            <!----><span>markAllAsDone</span></button> <button disabled="disabled" type="button" class="el-button el-button--primary is-disabled">
            <!---->
            <!----><span>reset</span></button> <button disabled="disabled" type="button" class="el-button el-button--primary is-disabled">
            <!---->
            <!----><span>save</span></button></div>
        <hr>
        <ul class="items">
          <li>
            <div class="el-row is-justify-space-between el-row--flex">
              <div class="el-col el-col-22">XXX</div>
              <div class="el-col el-col-2"><button type="button" class="el-button el-button--default el-button--mini">
                  <!----><i class="el-icon-check"></i>
                  <!----></button></div>
            </div>
          </li>
        </ul>
        <div class="footer">itemsLeft</div>
      </div>
    </div>
    <div class="el-col el-col-12">
      <div class="tasklist">
        <h1>alreadyDone</h1>
        <ul class="items"></ul>
      </div>
    </div>
  </div>
</div>
`;

exports[`tasklist app removes error class on valid input 1`] = `
<div>
  <div class="el-row">
    <div class="el-col el-col-12">
      <div class="tasklist">
        <h1>tasks</h1>
        <div class="el-input input"><input type="text" placeholder="addTask" value="" class="el-input__inner error"></div>
        <div class="el-button-group"><button disabled="disabled" type="button" class="el-button el-button--primary is-disabled">
            <!---->
            <!----><span>markAllAsDone</span></button> <button disabled="disabled" type="button" class="el-button el-button--primary is-disabled">
            <!---->
            <!----><span>reset</span></button> <button disabled="disabled" type="button" class="el-button el-button--primary is-disabled">
            <!---->
            <!----><span>save</span></button></div>
        <hr>
        <ul class="items"></ul>
        <div class="footer">itemsLeft</div>
      </div>
    </div>
    <div class="el-col el-col-12">
      <div class="tasklist">
        <h1>alreadyDone</h1>
        <ul class="items"></ul>
      </div>
    </div>
  </div>
</div>
`;

exports[`tasklist app removes error class on valid input 2`] = `
<div>
  <div class="el-row">
    <div class="el-col el-col-12">
      <div class="tasklist">
        <h1>tasks</h1>
        <div class="el-input input"><input type="text" placeholder="addTask" value="1234" class="el-input__inner"></div>
        <div class="el-button-group"><button disabled="disabled" type="button" class="el-button el-button--primary is-disabled">
            <!---->
            <!----><span>markAllAsDone</span></button> <button disabled="disabled" type="button" class="el-button el-button--primary is-disabled">
            <!---->
            <!----><span>reset</span></button> <button disabled="disabled" type="button" class="el-button el-button--primary is-disabled">
            <!---->
            <!----><span>save</span></button></div>
        <hr>
        <ul class="items"></ul>
        <div class="footer">itemsLeft</div>
      </div>
    </div>
    <div class="el-col el-col-12">
      <div class="tasklist">
        <h1>alreadyDone</h1>
        <ul class="items"></ul>
      </div>
    </div>
  </div>
</div>
`;
